pennylane/tape/tapes/qubit_param_shift.py

```diff
@@ -169,7 +169,13 @@ def processing_fn(results):
                 array[float]: 1-dimensional array of length determined by the tape output
                 measurement statistics
             """
-            return np.dot(coeffs, np.squeeze(results))
+            results = np.squeeze(results)
+
+            if results.dtype is not np.dtype("O"):
+                dot = lambda x: np.dot(coeffs, x)
+                return np.apply_along_axis(dot, 0, results)
+
+            return sum([c * r for c, r in zip(coeffs, results)])

```

Comment:
Feels safer to see this done by hand ;)

I had so many problems with unexpected behaviour of `dot` before, it is not trivial that one...

Reply:
What I am confused about, if there is a ragged array (for example made from `[[1] , [2, 3]]`), would the first entry only be multiplied by the first coefficient? Isn't this a strong assumption? Where does such a case appear?

Reply:
> Isn't this a strong assumption? Where does such a case appear?

Consider a QNode of the form:

```python
return qml.expval(qml.PauliZ(0)), qml.probs(wires=[1, 2])
```

If we have lots of tapes with this return (e.g., for parameter-shift), the results will look like this:

```python
[
    [[0.654], [0.123, 0.654, 0.123, 0.645]], # first tape result
    [[0.654], [0.123, 0.654, 0.123, 0.645]], # second tape result
    [[0.654], [0.123, 0.654, 0.123, 0.645]], # third tape result
    ....
]
```

(numbers randomly chosen!). As a result, `np.dot(coeff, results)` will not do what we want, but the list comprehension will.